### ISSUE TITLE
React native support 

### ISSUE THREAD

**Issue by Ashraf-Ali-aa** [electric-sql/pglite#87]
Hi I was wondering if this supported react native

**Comment by sqpollen** [comment]
Any updates on this?

**Comment by chris-garrett** [comment]
I'm very interested in this and can help if I can. I have no idea what is required and I don't know much about WASM but I am willing to dig in.

**Comment by Ashraf-Ali-aa** [comment]
https://github.com/cawfree/react-native-webassembly

looks like react native can support web assembly

**Comment by samwillis** [comment]
Hey,

We're looking into that (https://github.com/cawfree/react-native-webassembly) route to RN support, but it's going to require quite a bit of work as it wouldn't be compatible with the current build.

We are exploring two opens:

1. WASM with something like above or another third party runtime integration.
2. Having a native build of PGlite that can link to the native app and be wrapped with the TypeScript api.

Both of these will be dependent on the "libpglite" we are currently working on.

We hope to have a better idea soon, but rest assured, React native support is at the top of our to-do list.

**Comment by chris-garrett** [comment]
Seems like option 2 would make sense. Can we help with this? I don't see any obvious branches here.

**Comment by evelant** [comment]
The way to do it would be option 2, using [turbo native modules](https://github.com/reactwg/react-native-new-architecture/blob/main/docs/turbo-modules.md) and [cross platform turbo native modules](https://github.com/reactwg/react-native-new-architecture/blob/main/docs/turbo-modules-xplat.md). It may be as simple as plugging `PGliteInterface` into react native's codegen and adding boilerplate to call the native pglite functions from the generated interface.

**Comment by pmp-p** [comment]
> Seems like option 2 would make sense. Can we help with this? I don't see any obvious branches here.

@chris-garrett most likely we will be using WASI build of current code with experimental changes to wasi-sdk 23  to be tested with option 1, after success then we will just retarget assembly to C89 to make native builds for 2).

so far only blocker is wasm extensions loading which is not obvious on wasm3 ( or any wasi runtime so far)

**Comment by thruflo** [comment]
@samwillis is this the best place to track support for Expo and React Native? I assume it is.

**Comment by evelant** [comment]
Any progress on getting pglite running on react-native? A status update would be appreciated or some pointers to areas where help is needed.

**Comment by samwillis** [comment]
Hey @evelant 

We're making good progress, RN support depends on our "libpglite" project. This is going well, we have a good understanding of what's needs and have some early prototyping that showing that it's feasible.

we have two priorities right now: pg_dump (prerequisite to any our PG version bump) and RN support.

There's a chance of things to do to get there, but we have a firm plan: libpglite -> retarget to native C -> Kotlin and Swift wrappers -> React native.

**Comment by Andarius** [comment]
> Hey @evelant
> 
> We're making good progress, RN support depends on our "libpglite" project. This is going well, we have a good understanding of what's needs and have some early prototyping that showing that it's feasible.
> 
> we have two priorities right now: pg_dump (prerequisite to any PG version bump) and RN support.
> 
> There's a chain of things to do to get there, but we have a firm plan: libpglite -> retarget to native C -> Kotlin and Swift wrappers -> React native.

Did you consider using Nitro ? https://github.com/mrousavy/nitro

**Comment by rikur** [comment]
Collab with Expo? 👀

**Comment by niraj-khatiwada** [comment]
Waiting for React Native support as well.

**Comment by evelant** [comment]
I wish I could help push this forward but unfortunately I have little experience with C/C++ and time is too tight for me to spend time learning it at the moment ☹️ . I love what electric is building and am very excited to try it out and contribute in the future, IMO it's definitely the right direction to finally have a robust offline capable sync for apps!

**Comment by satvikpendem** [comment]
I assume this would also help with Flutter and other native platform support too right, if the pure C code can be interacted with via FFI? Would a WASM build also be feasible or just focusing on the pure C build for now?

**Comment by umsurething** [comment]
This will be an awesome addition and I am much more inclined to buy in to the ecosystem if it lands some time. Props for pglite btw. Incredible!!!

**Comment by grinono** [comment]
would be great to be able to use PGlite in react native.

**Comment by pmp-p** [comment]
| Due to iOS JIT restrictions 

@evelant True, wasi cannot help **directly** on iOS .. But if you give up on most contrib extensions there are ways to have wasi code running natively on a statically linked iOS app.

But since it is a tech i previously only used on android games, i whould just need to have a react iOS skeleton project that would call pglite typescript functions "as if they were native in C" ( ie stubs for me to replace)

**Comment by evelant** [comment]
https://nitro.margelo.com/docs/nitro-modules might help. Looks like a faster, easier to maintain, and overall better solution for including cross platform cpp in react native projects compared to expo modules or turbomodules.

**Comment by evelant** [comment]
Looking at how pglite builds unfortuantely I'm not sure there's much I can do to help -- the build process appears quite complex and I have little experience with make/cmake and bash scripts. Since it's all C however theoretically as far as I know there shouldn't be anything in the way of defining a nitro module interface and directly calling the C functions as needed. I just lack the expertise using C build tools to try it myself.

**Comment by copiltembel** [comment]
As @samwillis said earlier, we have a pretty good idea on how to achieve this - and more! We just need some more time to implement it.

**Comment by octalpixel** [comment]
> As [@samwillis](https://github.com/samwillis) said earlier, we have a pretty good idea on how to achieve this - and more! We just need some more time to implement it.

Do you think would be possible with https://github.com/callstackincubator/polygen from the callstack guys

**Comment by samwillis** [comment]
Hey @octalpixel 

Polygen is very cool! I've been keeping my eye on it. One of the routes we've been exploring is very similar, it's also using wasm to c. Using polygen may well work in future, but it's missing some critical functionality at the moment.

**Comment by octalpixel** [comment]
> Hey [@octalpixel](https://github.com/octalpixel)
> 
> Polygen is very cool! I've been keeping my eye on it. One of the routes we've been exploring is very similar, it's also using wasm to c. Using polygen may well work in future, but it's missing some critical functionality at the moment.

Super! Looking forward to it!

**Comment by stephancill** [comment]
Any update on the progress here or somewhere to track progress more granularly?

**Comment by bishoyroufael** [comment]
Any updates? Is there a way to make this work with Expo Web?

**Comment by evelant** [comment]
I'm really excited about the possibility of using pglite on react native. I wrote this offline-capable sync engine using pglite and electric but can't use it in my react native app yet! https://github.com/evelant/synchrotron

**Comment by ghandic** [comment]
Bumping @stephancill 's question, are the rough timelines or places anyone can help out on to achieve this?

**Comment by kaizi-ai** [comment]
any update on this? I am about to work on sth good on it.

**Comment by akodkod** [comment]
+1 for this, please 🙏

**Comment by e1nsteindotdev** [comment]
eagerly waiting

**Comment by evelant** [comment]
@samwillis We're still super keen on getting pglite in react-native, would you please let us know where things are at and if we can help with any of it? I wrote an offline sync engine with electric + pglite but can't adopt it yet and move it forward because our primary app is react native! https://github.com/evelant/synchrotron

**Comment by forabi** [comment]
I gave up on pglite. Even in browser, it's still not very mature and I have seen lots of crashes and ambiguous, hard-to-debug errors. I built an entire set of packages which allow electric to be used with wa-sqlite. It's been working great so far, aside from a few bugs. Eventually, I'll also make a RN version with op-sqlite and publish all these packages as open source.

**Comment by pmp-p** [comment]
> lots of crashes and ambiguous, hard-to-debug errors 

Though the same could be said about react native when ~~building~~ running native extensions.

I think a more detailed report could be usefull there, don't you ?

**Comment by evelant** [comment]
I gave it a shot and got the basics running on android. Hopefully this serves as a good starting point for official react native support. My PR is very proof-of-concept, I'm sure there's a lot broken/wrong but it shows basic db functionality on react-native android and lays all the necessary groundwork for full functionality. https://github.com/electric-sql/pglite/pull/774

**Comment by evelant** [comment]
Got it building and running on iOS too. Pglite on mobile seems like it can definitely work!

**Comment by roo12312** [comment]
Any updates?

**Comment by MrLoh** [comment]
Stop spamming. I'm sure there will be an update here when there are any news.

**Comment by evelant** [comment]
This is waiting on some other refactors in pglite. Then I will pick it up again. Anybody who's eager is welcome to work on improving it @roo12312 @niraj-khatiwada !

**Comment by tyrauber** [comment]
@evelant, dude, nice work! 

Took a few hours to figure out how to pull it all together, but this builds!  You are a champ!

```
› Logs for your project will appear below. Press Ctrl+C to exit.
iOS Bundled 335ms apps/example/index.ts (701 modules)
 LOG  [PGL] App mounted
 LOG  [PGL] Creating PGlite instance...
 LOG  [PGL] PGlite instance created
 LOG  [PGL] Waiting for ready...
 LOG  [PGL] Database ready!
 LOG  [PGL] Running SELECT 1...
 LOG  [PGL] Result: {"rows":[{"n":1}],"fields":[{"name":"n","dataTypeID":23}],"affectedRows":0}
[example.debug.dylib] [PGLiteNative] DEBUG: About to log 'About to call pgl_backend()'
 LOG  [PGL] Starting compatibility test suite...
 LOG  [PGL] ✅ Basic SQL: SELECT literal
 LOG  [PGL] ✅ Basic SQL: SELECT string literal
 LOG  [PGL] ✅ Basic SQL: SELECT multiple columns
 LOG  [PGL] ✅ Basic SQL: SELECT with arithmetic
 LOG  [PGL] ✅ Basic SQL: SELECT NULL handling
 LOG  [PGL] ✅ DDL: CREATE TABLE basic
 LOG  [PGL] ✅ DDL: CREATE TABLE with PRIMARY KEY
 LOG  [PGL] ✅ DDL: CREATE TABLE with SERIAL
 LOG  [PGL] ✅ DDL: CREATE TABLE with constraints
 LOG  [PGL] ✅ DDL: CREATE TABLE IF NOT EXISTS
 LOG  [PGL] ✅ DDL: ALTER TABLE ADD COLUMN
 LOG  [PGL] ✅ DDL: ALTER TABLE DROP COLUMN
 LOG  [PGL] ✅ DDL: DROP TABLE
 LOG  [PGL] ✅ DDL: DROP TABLE IF EXISTS
 LOG  [PGL] ✅ Data Types: INTEGER type
 LOG  [PGL] ✅ Data Types: BIGINT type
 LOG  [PGL] ✅ Data Types: TEXT type
 LOG  [PGL] ✅ Data Types: BOOLEAN type
 LOG  [PGL] ✅ Data Types: REAL/FLOAT type
 LOG  [PGL] ✅ Data Types: NUMERIC/DECIMAL type
 LOG  [PGL] ✅ Data Types: UUID type
 LOG  [PGL] ✅ Data Types: TIMESTAMP type
 LOG  [PGL] ✅ Data Types: TIMESTAMPTZ type
 LOG  [PGL] ✅ Data Types: DATE type
 LOG  [PGL] ✅ Data Types: JSON type
 LOG  [PGL] ✅ Data Types: JSONB type
 LOG  [PGL] ✅ Data Types: BYTEA type
 LOG  [PGL] ✅ Data Types: ARRAY type
 LOG  [PGL] ✅ DML: INSERT single row
 LOG  [PGL] ✅ DML: INSERT multiple rows
 LOG  [PGL] ✅ DML: INSERT RETURNING
 LOG  [PGL] ✅ DML: UPDATE single row
 LOG  [PGL] ✅ DML: UPDATE multiple rows
 LOG  [PGL] ✅ DML: DELETE single row
 LOG  [PGL] ✅ DML: DELETE all rows
 LOG  [PGL] ✅ DML: TRUNCATE
 LOG  [PGL] ✅ Queries: SELECT with WHERE
 LOG  [PGL] ✅ Queries: SELECT with ORDER BY
 LOG  [PGL] ✅ Queries: SELECT with LIMIT
 LOG  [PGL] ✅ Queries: SELECT with OFFSET
 LOG  [PGL] ✅ Queries: SELECT DISTINCT
 LOG  [PGL] ✅ Queries: SELECT with GROUP BY
 LOG  [PGL] ✅ Queries: SELECT with HAVING
 LOG  [PGL] ✅ Joins: INNER JOIN
 LOG  [PGL] ✅ Joins: LEFT JOIN
 LOG  [PGL] ✅ Aggregates: COUNT aggregate
 LOG  [PGL] ✅ Aggregates: SUM aggregate
 LOG  [PGL] ✅ Aggregates: AVG aggregate
 LOG  [PGL] ✅ Aggregates: MIN/MAX aggregates
 LOG  [PGL] ✅ Transactions: BEGIN/COMMIT transaction
 LOG  [PGL] ✅ Transactions: ROLLBACK transaction
 LOG  [PGL] ✅ Transactions: SAVEPOINT
 LOG  [PGL] ✅ Indexes: CREATE INDEX
 LOG  [PGL] ✅ Indexes: CREATE UNIQUE INDEX
 LOG  [PGL] ✅ Indexes: DROP INDEX
 LOG  [PGL] ✅ Constraints: FOREIGN KEY constraint
 LOG  [PGL] ❌ Constraints: CHECK constraint
 LOG  [PGL]    Error: CHECK constraint not enforced
 LOG  [PGL] ✅ JSON: JSON path extraction (->)
 LOG  [PGL] ✅ JSON: JSONB path extraction (->>)
 LOG  [PGL] ✅ JSON: JSONB containment (@>)
 LOG  [PGL] ✅ Functions: CONCAT function
 LOG  [PGL] ✅ Functions: LOWER/UPPER functions
 LOG  [PGL] ✅ Functions: LENGTH function
 LOG  [PGL] ✅ Functions: SUBSTRING function
 LOG  [PGL] ✅ Functions: TRIM function
 LOG  [PGL] ✅ Date/Time: NOW() function
 LOG  [PGL] ✅ Date/Time: CURRENT_DATE
 LOG  [PGL] ✅ Date/Time: EXTRACT function
 LOG  [PGL] ✅ Date/Time: DATE_TRUNC function
 LOG  [PGL] ✅ Subqueries: Scalar subquery
 LOG  [PGL] ✅ Subqueries: IN subquery
 LOG  [PGL] ✅ Subqueries: EXISTS subquery
 LOG  [PGL] ✅ CTEs: Basic CTE (WITH clause)
 LOG  [PGL] ✅ Expressions: CASE expression
 LOG  [PGL] ✅ Expressions: COALESCE function
 LOG  [PGL] ✅ Expressions: NULLIF function
 LOG  [PGL] ✅ Parameters: Parameterized query ($1)
 LOG  [PGL] ✅ Parameters: Multiple parameters ($1, $2)
 LOG  [PGL] ✅ UPSERT: INSERT ON CONFLICT DO NOTHING
 LOG  [PGL] ✅ UPSERT: INSERT ON CONFLICT DO UPDATE
 LOG  [PGL] ✅ Views: CREATE VIEW
 LOG  [PGL] ✅ Functions: generate_series function
 LOG  [PGL] 
=== SUMMARY ===
 LOG  [PGL] Total: 82
 LOG  [PGL] Passed: 81
 LOG  [PGL] Failed: 1
 LOG  [PGL] Skipped: 0
 LOG  [PGL] Pass Rate: 98.8%
```

**Comment by evelant** [comment]
It's definitely NOT ready to be used for anything -- there are certainly a ton of bugs. It's on hold right now while the electric-sql team are working to refine how the pglite build works. After that I'll probably take another pass at it to see if it can be made fully usable (I don't see why it's not possible given the proof of concept here!)

**Comment by tyrauber** [comment]
Yeah, definitely not production ready - but close (IMO).  

The big issue I have had in development is initialization crashes around native loading of data, native classes and symbols. I found bundling resources is the best approach to solve flattening of resources which leads to the crash scenario. Most of the work I did was on bash build scripts to get everything properly compiled.   (Where is the work on build scripts happening?).

I did make a few upstream changes which I haven't PR'd yet. These build on your work. Should I PR to your remote?

[postgres-pglite](https://github.com/tyrauber/postgres-pglite):
- Fix errno macro expansion issue on all Apple platforms [aa8721a](https://github.com/tyrauber/postgres-pglite/commit/aa8721a461497df00d9af696076258948472c85e)
- feat: Add static extension support for React Native mobile builds [caa9e41](https://github.com/tyrauber/postgres-pglite/commit/caa9e4186473261ddf97adbf065398ea23b2b282)

**Comment by bendytree** [comment]
fwiw - I needed native pglite on ios too, so I had Fable build/test it baed off this repo.

> [pglite-ios](https://github.com/bendytree/pglite-ios)

It takes the WASI build of PGlite (thanks @pmp-p and @moznion), and compiles it as a normal iOS library. No wasm runtime, no interpreter, safe for app store. pgvector is built in. It include an optional local server, so anything with a Postgres driver (e.g. js in an RN app) can connect. I needed pgvector so it's built in.

 Real-world numbers from my iPhone 17 Pro:
 - Database opens in 0.3s (ships pre-initialized)
 - Simple queries well under a millisecond
 - Full test suite (transactions, error recovery, crash recovery/persistence, vector indexes) passes on macOS, simulator, and device
 - 3MB binary, starts at 27 MB on iOS disk, 40MB RAM
